### ISSUE TITLE
Maybe the GFW can identification the MTProxy

### ISSUE THREAD

**Issue by DavidDen** [TelegramMessenger/MTProxy#278]
When I build up a MTProxy, the GFW will block the MTProxy port very quick.

**Comment by DavidDen** [comment]
If I change a port, it can use about 30 mins

**Comment by imliuruiqi** [comment]
Mine worked for about 1day.... Then it can't be connected.

**Comment by jazhLeo** [comment]
what am i, just 1 second....

**Comment by cpdyj** [comment]
After 12:00 today, my all proxy (include my self server) has been blocked by GFW.  
SSH connect ok, only MTProxy has been blocked.

**Comment by number201724** [comment]
Similarly, I am in Shenzhen and it is banned within 30 days.

**Comment by jmluang** [comment]
blocked suddenly. change port not work.

**Comment by number201724** [comment]
> blocked suddenly. change port not work.

No, I am different from yours. After I was banned, even if I used telnet to test, I could not connect, and the entire port was disabled.

**Comment by Baz2013** [comment]
me too, I changed a port , it worked less than five minutes.

**Comment by summershrimp** [comment]
pass `-v` param and paste the log here? Mine shows 
```
[2747][2019-03-22 16:07:26.838420 local] New connection xxxx ->xxxx
[2747][2019-03-22 16:07:29.766274 local] trying to determine connection type
[2747][2019-03-22 16:07:29.767066 local] invalid "random" 64-byte header, entering global skip mode
[2747][2019-03-22 16:07:30.938654 local] New connection xxxx -> xxxx
[2747][2019-03-22 16:07:31.711521 local] socket 79: disconnected (epoll_ready=2005), cleaning
[2747][2019-03-22 16:07:31.712199 local] Closing connection socket #79
[2747][2019-03-22 16:07:31.720602 local] socket 80: disconnected (epoll_ready=2005), cleaning
[2747][2019-03-22 16:07:31.721360 local] Closing connection socket #80
[2747][2019-03-22 16:07:32.877193 local] socket 82: disconnected (epoll_ready=2005), cleaning
[2747][2019-03-22 16:07:32.877631 local] socket 81: disconnected (epoll_ready=2005), cleaning
[2747][2019-03-22 16:07:32.877913 local] Closing connection socket #82
[2747][2019-03-22 16:07:32.878143 local] Closing connection socket #81
```

**Comment by Yesterday17** [comment]
IPv6 works well. Only IPv4 is blocked. I'm using IPv6 address instead now but I don't have IPv6 all the time.(cry)

**Comment by SheffieldLee** [comment]
Me too

**Comment by imliuruiqi** [comment]
port unblocking after about 2days

**Comment by Yesterday17** [comment]
I'm using port forwarding to solve the problem. I ported port 3000-5000 to the origin port, and when a port is blocked, i can use another port instead(until the ip is banned).

**Comment by lut777** [comment]
same problem.
The IP could be pinged, but the port could not be curled. 
obviously, the port has been banned by GFW.

**Comment by suncunhu** [comment]
fxxk, gfw
On PC, I can use local socks5 replace MTProxy, avoid change vps MTProxy port frequently
However, It's a little difficult On Mobile Phone, because I'm uncertain local socks5 port, so can't do this way

**Comment by fa1247** [comment]
Me too. If I open MTproxy on a random port, it will be blocked in minutes, or worse, in seconds. Maybe MTProto has some obvious features that make it easy to be detected. We have been forced to give up using mtproto. Please do something. Thanks.

**Comment by SheffieldLee** [comment]
I can connect again, how about you?

**Comment by halhaling** [comment]
Same to me. Socks running well but MTProxy doesn't work sooner after I change the port. What a pity.

**Comment by Joeyxworks** [comment]
Same. Have anyone tried domain name instead?

**Comment by fa1247** [comment]
> Same. Have anyone tried domain name instead?

I have tried. Same.

**Comment by liaogang** [comment]
+1.  Now I'm using shadowsocks instead. iOS 11.

**Comment by 494310051** [comment]
same as me.
I have 8 server of MTproxy & 4-6 server's prot blocked every day

**Comment by jay816314** [comment]
Same. Use brook/v2ray instead.

**Comment by CicoZhang** [comment]
Thank you guys for sharing this information. I am not alone. My mom complained to me about this and now I am sure about the cause.

**Comment by 494310051** [comment]
> Same. Use brook/v2ray instead.

use v2ray/Brook work as MTproxy inner telegram?

**Comment by jay816314** [comment]
> > Same. Use brook/v2ray instead.
> 
> use v2ray/Brook work as MTproxy inner telegram?

No, it's another application. Proxy inner telegram only support MTproxy and SOCKS5.
You can try using SOCKS5, it works for me now.

**Comment by 494310051** [comment]
> > > Same. Use brook/v2ray instead.
> > 
> > 
> > use v2ray/Brook work as MTproxy inner telegram?
> 
> No, it's another application. Proxy inner telegram only support MTproxy and SOCKS5.
> You can try using SOCKS5, it works for me now.

yes,it work well on pc , but ... mobile...

**Comment by jay816314** [comment]
> 
> yes,it work well on pc , but ... mobile...
SOCKS5 work on my mobile(iOS)。

**Comment by gouff** [comment]
This tool should be upgrade to face the upgraded GFW.

**Comment by wi1dcard** [comment]
目前我探索的解决方案有三：

1. 使用 iptables + ss-redir + Python 版本的 MTProto Proxy，将 MTProto 的流量转发给 ss-redir：<https://lifetyper.com/posts/use_ssr_to_redirect_mtproto_proxy_inside_gfw.html>  （官方 C 语言版本的 MTProto Proxy 使用 iptables NAT redirect 似乎无效，水平有限，原因暂时未知）

2. 国内搭建 Socks5 代理，使用 iptables 将目标 IP 为 [AS62041](https://ipinfo.io/AS62041)、[AS59930](https://ipinfo.io/AS59930) 全部转发到 ss-redir；对于其它目标 IP 段的流量留在国内，以防被滥用。 

3. 使用 V2ray + 国内机器转发 MTProto 流量到 VMess / Shadowsocks: <https://steemit.com/cn/@v2ray/v2ray-mtproto> 

以上三个方案效果如下，延迟差距不大。

![image](https://user-images.githubusercontent.com/25447002/56495475-69649a80-64e5-11e9-9f49-7813cffda9f0.png)

**Comment by sqliuchang** [comment]
> 目前我探索的解决方案有三，都是通过国内机器做跳板：
> 
> 1. 使用国内机器 + iptables + ss-redir + Python 版本 MTProto Proxy，将 MTProto 的流量转发给 ss-redir：https://lifetyper.com/posts/use_ssr_to_redirect_mtproto_proxy_inside_gfw.html  （官方 C 语言版本的 MTProto Proxy 使用 iptables NAT redirect 似乎无效，水平有限，原因暂时未知）
> 2. 国内搭建 Socks5 代理，使用 iptables 将目标 IP 为 [AS62041](https://ipinfo.io/AS62041)、[AS59930](https://ipinfo.io/AS59930) 全部转发到 ss-redir；对于其它目标 IP 段的流量留在国内，以防被滥用。
> 3. 国内搭建 V2ray 转发 MTProto 流量到 VMess / Shadowsocks: https://steemit.com/cn/@v2ray/v2ray-mtproto
> 
其实还有一种，就是套一层kcptun，类似于V2ray转发。
目前国内跳板我用的是腾讯云学生机，1M的带宽聊天还算能用。还有就是家里的路由器，但是开着kcptun cpu负载略大，作为备用方案。

**Comment by wi1dcard** [comment]
@sqliuchang 我是考虑到给女票用，所以需要满足两个条件：

1. 4G 也能用
2. 不要求手机上安装 SS 等第三方软件

其实如果家里路由器可以玩的话，就有很多方式解决了，不用 kcptun 也可以，直接用 SS 也没问题。另外，kcptun 基于 Golang，不知道是不是我的姿势不对，总感觉 Golang 对于 ARM / Mips 的优化似乎真的不太好，资源占用很高，可能 Golang 的初衷也不是针对这些小型设备的吧。另外 KCP 也是真真的烧性能... GCP 每月 5 刀的实例只能跑到 20Mbps 左右 CPU 就爆满了。

**Comment by sqliuchang** [comment]
> @sqliuchang 我是考虑到给女票用，所以需要满足两个条件：
> 
> 1. 4G 也能用
> 2. 不要求手机上安装 SS 等第三方软件


我傻了，我和你一模一样的情况，我也是考虑怎么让女票用起来方便。。。
最终我还是给她装了shadowrocket，添加好节点和场景，让她4G开代理，自己路由器直连。


> 其实如果家里路由器可以玩的话，就有很多方式解决了，不用 kcptun 也可以，直接用 SS 也没问题。

有公网ip的宽带在老家，所以老家的路由器基本上是当成国内跳板来使用的。主要还是考虑在外面的时候手机怎么用tg。


> 另外，kcptun 基于 Golang，不知道是不是我的姿势不对，总感觉 Golang 对于 ARM / Mips 的优化似乎真的不太好，资源占用很高，可能 Golang 的初衷也不是针对这些小型设备的吧。另外 KCP 也是真真的烧性能... GCP 每月 5 刀的实例只能跑到 20Mbps 左右 CPU 就爆满了。

是的，我老家的K3路由器开在normal模式平均负载提高0.7左右，fast2就更高了，还什么都没干CPU就快满载了。。

**Comment by wi1dcard** [comment]
@sqliuchang Shadowrocket 确实不错，我之前也在用，不过现在转投 Quantummult 了，路由规则很强大。

另外，我也是买的软路由，不过没有上虚拟机、Koolshare 改版 OpenWrt 之类的，是自己直接配置的 Ubuntu Server，感觉比较干净。看起来渣渣 CPU 的 x86 机器都能随便秒掉 Mips / ARM，我家百兆宽带，重度 SS 用户，CPU 和内存几乎一直维持在很低的状态：

![image](https://user-images.githubusercontent.com/25447002/57296885-ac875600-7100-11e9-9165-8924f68a3e38.png)

**Comment by rkline0x** [comment]
This thread is seven years old and the core issue hasn't changed — the obfuscated2
protocol is fundamentally fingerprintable. It has no TLS framing, so DPI doesn't even
need deep analysis: the traffic simply doesn't match any known protocol.

I've been working on this in [teleproxy](https://github.com/teleproxy/teleproxy), a
maintained fork. The fake-TLS transport wraps connections in real TLS 1.3 records with a
Chrome-matching ClientHello (correct JA3 hash, GREASE extensions per RFC 8701). Server
responses are fragmented across separate TCP segments with ±32 bytes of size variation
per connection, so DPI can't pattern-match the handshake structure.

Dynamic Record Sizing varies TLS record lengths the way Cloudflare/Caddy do in practice
(1450 → 4096 → 16144 bytes with random noise), defeating statistical traffic analysis.
You can also put nginx with a real certificate behind teleproxy — active probes that send
an invalid secret get forwarded to the real site instead of getting a connection reset.

All of this is transparent to the Telegram app — no client changes needed, just an `ee`
prefix secret instead of `dd`.
